### ISSUE TITLE
Opportunity: operator precedence table

### ISSUE THREAD

**Issue by mikez** [stedolan/jq#2425]
Does anyone here think operator precedence tables are helpful? I do. :) (Examples: [Python](https://docs.python.org/3/reference/expressions.html#operator-precedence) and [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table).)

Is there already such a table for jq? If not, would it make sense to include it in the [Wiki](https://github.com/stedolan/jq/wiki) or elsewhere?

## Example table (feel free to copy and change)

Here's a first draft based on [Rosetta Code](https://rosettacode.org/wiki/Operator_precedence#jq), [parser.y](https://github.com/stedolan/jq/blob/master/src/parser.y), some simplifying assumptions, and guesses. Somebody who knows the codebase better than me, would be a good candidate to correct and extend this. :)

Order: from highest precedence (most binding) to lowest precedence (least binding).

| Operator                                            | Description                      |
| --------------------------------------------------- | -------------------------------- |
| `x?`                                                | Error Suppression                |
| `-x`                                                | Negative                         |
| `*`, `/`, `%`                                       | Multiplication, Division, Modulo |
| `+`, `-`                                            | Addition, Subtraction            |
| `==`, `!=`, `<`, `>`,`<=`, `>=`                     | Comparisons                      |
| `and`                                               | Boolean AND                      |
| `or`                                                | Boolean OR                       |
| `=`, <code>\|=</code>, `+=`, `-=`, `*=`, `/=`, `%=` | Update-assignment                |
| `//`                                                | Alternative                      |
| `,`                                                 | Comma                            |
| <code>\|</code>                                     | Pipe                             |
| `label $variable`                                   | Labels                           |
| `try … catch …`                                     | Try expression                   |
| `if … then … end`                                   | Conditional expression           |
| `foreach … as … (…)`                                | Loop expression                  |
| `reduce … as … (…)`                                 | Reduce expression                |
| `… as $variable`                                    | Variable definition expression   |
| `def … ; …`                                         | Function expression              |

**Comment by wader** [comment]
Good idea i think. Had a look at the wiki and found this https://github.com/stedolan/jq/wiki/jq-Language-Description#operators-priority, maybe it can be extended?

**Comment by mikez** [comment]
@wader Good find!

It seems @fadado added this in 2019 with the comment _"Added operators priority table. Please, check the table validity, specially the new added operator `?//`"_ and seemingly it hasn't been revised since. So maybe there was some uncertainty there too? :) Or maybe the concept of operator precedence is less applicable in the world of jq?

Some proposals:

1. Find someone to verify the correctness of this.
2. Change the section title to _Operator Precedence_ for consistency with other languages and ease it being found.
3. Add clickable links akin to the [JavaScript table](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table).

**Comment by x-yuri** [comment]
Is it asc or desc? I'd argue that usually it's desc. Anyway the direction should be specified.

**Comment by emanuele6** [comment]
> Order: from highest precedence (most binding) to lowest precedence (least binding).

**Comment by x-yuri** [comment]
Oh, indeed. I probably had the [wiki][a] in mind. Order is not specified there. And it's apparently from lowest to highest. Which I think is rather unusual. At least from my experience...

[a]: https://github.com/jqlang/jq/wiki/jq-Language-Description#operators-priority

**Comment by nicowilliams** [comment]
Thank you for this.

Yes, the wiki is currently the best place for this work, and you can go ahead and edit it.  Eventually we'd like to come up with a more formal specification of the language and put it up as a separate repository in the `jqlang` org.

**Comment by derelbenkoenig** [comment]
Hello, I came across this issue while trying to find the operator precedence rules myself, and while it is useful on the wiki, I would have expected it to be in the manual page as well. I have read the manual page many times but only just found out the wiki existed. Is there any reason not to add it there?
(Frustratingly, the manual does _mention_ precedence rules in the section on the `//` operator, but only to say that precedence rules can make it confusing, without actually telling you what they are)

**Comment by rgoldberg** [comment]
@itchyny @nicowilliams @wader Sorry to ping you, but is there any way for this to be documented?

It's difficult to use a language without knowing its operator precedence.

Per version documentation would be useful, as I am sticking with 1.7.1 compatibility, as it's the version supplied with recent macOS versions.

Thanks for any help you can provide & for your work on jq.

**Comment by mikez** [comment]
@rgoldberg Seeing it's been 4 years since I asked this, it might be too much work for one person to do. Maybe, we can break up the work, to make it more manageable. If you'd like to start, you could begin to document all operators which are currently missing from the table in the Wiki. Check out the [parser.y](https://github.com/stedolan/jq/blob/master/src/parser.y) file.

**Comment by zachary-krepelka** [comment]
For what it's worth, I thought this was interesting, and I wanted to share it
here.  Maybe it would be better as an entry in the cookbook on the wiki.

The [jq implementation of jq][1] has each binary operator as an object literal
in its source code with fields for precedence and associativity.  This will be
found ([here][2]) in `jqjq.jq` under the top-level function `parse/0`, under its
local function `_p/1`, under its local function `_op_prec_climb/2`, and under
its local function `_ops/0` as an if-elif-else switch statement, like this.

```jq
elif .pipe then           {prec: 0, name: "|",   assoc: "right"}
elif .comma then          {prec: 1, name: ",",   assoc: "right"}
elif .slash_slash then    {prec: 2, name: "//",  assoc: "right"}
elif .equal then          {prec: 3, name: "=",   assoc: "none"}
elif .pipe_equal then     {prec: 3, name: "|=",  assoc: "none"}
```

As a consequence of this, we can scrape the source code directly to generate an
operator precedence and associativity table on the fly.  For example, I have
this in my `~/.bashrc`.

```bash
function jq_op_prec_and_assoc_table {

  local OPTIND help_msg="usage: $FUNCNAME [-r]" opt_reverse=false

  while getopts :hr opt
  do
    case "$opt" in
      h) echo "$help_msg"; return 0;;
      r) opt_reverse=true;;
    esac
  done

  wget -qO- https://raw.githubusercontent.com/wader/jqjq/refs/heads/master/jqjq.jq \
    | sed -n \
      '
      /prec:/{
        s/.*{/{/
        s/\([a-z]\+\):/"\1":/g
        p
      }
      ' \
    | jq -rs \
      '
      {
        left: "left-to-right",
        right: "right-to-left",
        none: "none"
      } as $lookup
      | foreach group_by(.prec)[] as $level
        (
          null;
          $level[0] as $representative_operator
          | reduce $level[] as $operator
            (
              {
                precedence_level: $representative_operator.prec,
                associativity: $lookup[$representative_operator.assoc]
              };
              .operator_set += [$operator.name]
            )
        )
      | [.precedence_level, (.operator_set | join(" ")), .associativity]
      | @tsv
      ' \
    | if $opt_reverse
      then cat
      else tac
      fi \
    | column -t -s $'\t' -N PRECEDENCE,OPERATORS,ASSOCIATIVITY
}
```

Invoking this function by itself prints a table ordered from highest precedence
(most binding) to lowest precedence (least binding), as shown below.  Passing
the `-r` flag reverses the table so that it's ordered from lowest precedence
(least binding) to highest precedence (most binding).

```text
person@computer:~$ jq_op_prec_and_assoc_table
PRECEDENCE  OPERATORS            ASSOCIATIVITY
8           * / %                left-to-right
7           + -                  left-to-right
6           == != < <= > >=      none
5           and                  left-to-right
4           or                   left-to-right
3           = |= += -= *= /= %=  none
2           //                   right-to-left
1           ,                    right-to-left
0           |                    right-to-left
person@computer:~$
```

Of course, this table is limited in that 1) it is specific to jqjq and 2) it
only shows binary operators.

[1]: https://github.com/wader/jqjq
[2]: https://github.com/wader/jqjq/blob/644d9650359eaef9d82bdd2cb146a54e891a0538/jqjq.jq#L229-L256

**Comment by wader** [comment]
Nice. Note that `,`-associativity might be wrong in jqjq for the moment, there is a TODO about it. I don't remember the details now why it ended up that way. I did quick test changing it to left but one make test seems to get stuck in some recursion loop etc (`def f: 1; def g: f, def f: 2; def g: 3; f, def f: g; f, g; def f: 4; [f, def f: g; def g: 5; f, g]+[f,g]`), not sure if it due to AST ends up wrong or something else.

**Comment by 6494576** [comment]
刚给团队分享了，大家都在讨论 https://520menghuan.pages.dev/ 上那个日漫站和这个项目的架构差异，有懂哥来分析下吗？
